library/core/src/ptr/mut_ptr.rs

```diff
@@ -474,6 +475,21 @@ impl<T: ?Sized> *mut T {
     #[stable(feature = "pointer_byte_offsets", since = "1.75.0")]
     #[rustc_const_stable(feature = "const_pointer_byte_offsets", since = "1.75.0")]
     #[cfg_attr(miri, track_caller)] // even without panics, this helps for Miri backtraces
+    #[requires(
+        // If count is zero, any pointer is valid including null pointer.
+        (count == 0) || 
+        // Else if count is not zero, then ensure that subtracting `count` doesn't 
+        // cause overflow and that both pointers `self` and the result are in the 
+        // same allocation 
+        (mem::size_of_val_raw(self) != 0 &&
+            (self.addr() as isize).checked_add(count).is_some() && 
+            kani::mem::same_allocation(self as *const T, self.wrapping_byte_offset(count) as *const T))
```

Comment:
We should update all contracts to use `ub_checks` for `same_allocation`.

Reply:
Addressed in new commits.